numba/cuda/cudadrv/driver.py

```diff
@@ -2605,15 +2609,28 @@ def error_log(self):
     def add_ptx(self, ptx, name):
         """Add PTX source in a string to the link"""
 
+    @abstractmethod
+    def add_cu(self, cu, name):
+        """Add CUDA source in a string to the link. The name of the source
+        file should be specified in `name`."""
+
     @abstractmethod
     def add_file(self, path, kind):
         """Add code from a file to the link"""
 
+    def add_cu_file(self, path):
+        with open(path, 'rb') as f:
```

Comment:
I think it might as well just try, rather than reimplementing the error-handling logic of `open()`. For example, if you try to link a nonexistent file, the last line of the traceback is pretty clear:

```
FileNotFoundError: [Errno 2] No such file or directory: '/home/gmarkall/numbadev/numba/numba/cuda/tests/doc_examples/ffi/functioddns.cu'
```

Reply:
ok, let's leave it as-is. It does state the issue, which is probably sufficient to start a bug report for e.g. the maintainers of a library that is using this functionality.